File: csharp/BCRand/BCRand/SecureRandomNumberGenerator.cs
using System.Security.Cryptography;
using CryptoRng = System.Security.Cryptography.RandomNumberGenerator;

namespace BlockchainCommons.BCRand;

/// <summary>
/// A cryptographically secure random number generator.
/// Thread-safe; delegates to <see cref="System.Security.Cryptography.RandomNumberGenerator"/>.
/// </summary>
public sealed class SecureRandomNumberGenerator : IRandomNumberGenerator
{
    /// <summary>Shared thread-safe instance.</summary>
    public static readonly SecureRandomNumberGenerator Shared = new();

    public uint NextUInt32() => (uint)NextUInt64();

    public ulong NextUInt64()
    {
        Span<byte> buf = stackalloc byte[8];
        CryptoRng.Fill(buf);
        return BitConverter.ToUInt64(buf);
    }

    public byte[] RandomData(int size)
    {
        var data = new byte[size];
        CryptoRng.Fill(data);
        return data;
    }

    public void FillRandomData(Span<byte> data)
    {
        CryptoRng.Fill(data);
    }

    /// <summary>Generate cryptographically secure random bytes (module-level convenience).</summary>
    public static byte[] SecureRandomData(int size) => Shared.RandomData(size);

    /// <summary>Fill a buffer with cryptographically secure random bytes (module-level convenience).</summary>
    public static void SecureFillRandomData(Span<byte> data) => Shared.FillRandomData(data);
}

File: kotlin/bc-rand/src/main/kotlin/com/blockchaincommons/bcrand/SecureRandomNumberGenerator.kt
package com.blockchaincommons.bcrand

import java.security.SecureRandom

private val secureRandom = SecureRandom()

/**
 * Generates a byte array of cryptographically strong random bytes of the given size.
 *
 * @param size The number of random bytes to generate.
 * @return A new [ByteArray] containing [size] cryptographically strong random bytes.
 */
fun randomData(size: Int): ByteArray =
    ByteArray(size).also { secureRandom.nextBytes(it) }

/**
 * Fills the given byte array with cryptographically strong random bytes.
 *
 * @param data The byte array to fill with random bytes.
 */
fun fillRandomData(data: ByteArray) {
    secureRandom.nextBytes(data)
}

/**
 * A random number generator that can be used as a source of
 * cryptographically-strong randomness.
 *
 * Backed by [java.security.SecureRandom].
 */
class SecureRandomNumberGenerator : RandomNumberGenerator() {

    override fun nextU32(): UInt = secureRandom.nextInt().toUInt()

    override fun nextU64(): ULong = secureRandom.nextLong().toULong()

    override fun randomData(size: Int): ByteArray =
        ByteArray(size).also { secureRandom.nextBytes(it) }

    override fun fillRandomData(data: ByteArray) {
        secureRandom.nextBytes(data)
    }
}

fun secureRandomNumberGenerator(): SecureRandomNumberGenerator =
    SecureRandomNumberGenerator()
